hydromt/workflows/region.py

```diff
@@ -0,0 +1,293 @@
+        (x, y), or a region of interest (bounding box or geometry) for which the
+        basin IDs are looked up. The basins withint the area of interest can be further
+        filtered to only include basins with their outlet within the area of interest
+        ('outlets': true) of stream threshold arguments (e.g.: 'uparea': 1000).
+
+        Common use-cases include:
+
+        * {'basin': ID}
+
+        * {'basin': [ID1, ID2, ..]}
+
+        * {'basin': [x, y]}
+
+        * {'basin': [[x1, x2, ..], [y1, y2, ..]]}
+
+        * {'basin': /path/to/point_geometry}
+
+        * {'basin': [xmin, ymin, xmax, ymax]}
+
+        * {'basin': [xmin, ymin, xmax, ymax], 'outlets': true}
+
+        * {'basin': [xmin, ymin, xmax, ymax], '<variable>': threshold}
+
+        Subbasins are defined by its outlet locations and include all area upstream
+        from these points. The outlet locations can be passed as xy coordinate pairs,
+        but also derived from the most downstream cell(s) within a area of interest
+        defined by a bounding box or geometry, optionally refined by stream threshold
+        arguments.
+
+        The method can be speed up by providing an additional ``bounds`` argument which
+        should contain all upstream cell. If cells upstream of the subbasin are not
+        within the provide bounds a warning will be raised. Common use-cases include:
+
+        * {'subbasin': [x, y], '<variable>': threshold}
+
+        * {
+            'subbasin': [[x1, x2, ..], [y1, y2, ..]],
+            '<variable>': threshold, 'bounds': [xmin, ymin, xmax, ymax]
+            }
+
+        * {'subbasin': /path/to/point_geometry, '<variable>': threshold}
+
+        * {'subbasin': [xmin, ymin, xmax, ymax], '<variable>': threshold}
+
+        * {'subbasin': /path/to/polygon_geometry, '<variable>': threshold}
+
+        Interbasins are similar to subbasins but are bounded by a bounding box or
+        geometry and do not include all upstream area. Common use-cases include:
+
+        * {'interbasin': [xmin, ymin, xmax, ymax], '<variable>': threshold}
+
+        * {'interbasin': [xmin, ymin, xmax, ymax], 'xy': [x, y]}
+
+        * {'interbasin': /path/to/polygon_geometry, 'outlets': true}
+    """
+    kwargs = region.copy()
+    # NOTE: the order is important to prioritize the arguments
+    options = {
+        "basin": ["basid", "geom", "bbox", "xy"],
+        "subbasin": ["geom", "bbox", "xy"],
+        "interbasin": ["geom", "bbox", "xy"],
+        "geom": ["geom"],
+        "bbox": ["bbox"],
+        "mesh": ["UgridDataArray"],
+        "grid": ["raster"],
+    }
+    kind = next(iter(kwargs))  # first key of region
+    value0 = kwargs.pop(kind)
+
+    if kind in PLUGINS.model_plugins:
+        model_class = PLUGINS.model_plugins[kind]
+        other_model = model_class(root=value0, mode="r", logger=logger)
+        value0 = other_model.region
+        kwargs = dict(geom=other_model.region)
+        kind = "geom"
+    elif kind == "mesh":
```

Comment:
Should `parse_region("mesh")` be supported? Or is it something that `MeshComponent` should handle instead?

Reply:
As far as I can tell the "mesh" code path doesn't even return a `GeoDataframe`. if that is the case then I think it should either not be supported here or we should find a way to make it compatible with that return type

Reply:
And more changes... So actually before kinds model, grid and mesh actually did not return a geodataframe but the object you were asking for ie a grid, a mesh or the model. And the other kinds were also not returning a geodataframe but a dict and only geom option was returning a geodataframe in the dict....

So this is quite a big change. For most kinds a good change but then we have to deal with the other kinds like model, grid and mesh:

- model: as returning the entire model is quite specific for now to Delwaq, I agree to return the model.region and I'll deal with it in the delwaq plugin
- grid and mesh: I really think you should get a grid and mesh objects back here for it to be useful for plugins ie be able to re-use the schematisation and not just the bounding box. This said maybe some models only want it for the extent and want their own shematisation so it may be okay to just return the bounding box after all. But then we will have to duplicates some code again in the plugins if they do want to get the xarray object of gird and mesh... Maybe add two other functions like parse_region_grid and parse_region_mesh that would return xarray objects and in parse_region we can call then and then just create the geom from the bounding box?